Mathlib/Tactic/Widget/StringDiagram.lean

```diff
@@ -0,0 +1,317 @@
+/-
+Copyright (c) 2024 Yuma Mizuno. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yuma Mizuno
+-/
+import ProofWidgets.Component.PenroseDiagram
+import Mathlib.Tactic.CategoryTheory.Monoidal
+
+/-!
+# String Diagrams
+
+This file provides tactic/meta infrastructure for displaying string diagrams for morphisms
+in monoidal categories in the infoview.
+
+String diagrams are graphical representations of morphisms in monoidal categories, which are
+useful for rewriting computations. More precisely, objects in a monoidal category is represented
+by strings, and morphisms between two objects is represented by nodes connecting two strings
+associated with the objects. The tensor product `X ⊗ Y` corresponds to putting strings associated
+with `X` and `Y` horizontally (from left to right), and the composition of morphisms `f : X ⟶ Y`
+and `g : Y ⟶ Z` corresponds to connecting two nodes associated with `f` and `g` vertically (from
+top to bottom) by strings associated with `Y`.
+
+Currently, the string diagram widget provided in this file deals with equalities of morphisms
+in monoidal categories. It displays string diagrams corresponding to the morphisms for the
+left-hand and right-hand sides of the equality.
+
+Some examples can be found in `test/StringDiagram.lean`.
+
+When drawing string diagrams, it is common to ignore associators and unitors. We follow this
+convention. To do this, we need to extract non-structural morphisms that are not associators
+and unitors from lean expressions. This operation is performed using the `Tactic.Monoidal.eval`
+function.
+
+In this file, we will sometimes refer to objects and morphisms in monoidal categories
+as 1-morphisms and 2-morphisms, borrowing the terminology of the theory of 2-categories.
```

Comment:
This is a little confusing to me (but I am not an n-categorist)! Are objects not 0-morphisms, and morphisms 1-morphisms? Then an example of 2-morphisms would be the natural transformations in Cat.

Reply:
Thank you for the comment. My explanation here wasn't very helpful. I added an explanation about the relationship between monoidal categories and bicategories and why I used bicategory terminology in this file.